pkg/ovs/openflow/ofctrl_builder_test.go

```diff
@@ -15,30 +15,1952 @@
+			}
+
+			flowMod := getFlowMod(t, fb.Done())
+			assert.Equal(t, 1, len(flowMod.Match.Fields))
+			checkMatchField(t, tc.expectedMatchField, &flowMod.Match.Fields[0])
+			assert.Contains(t, getFlowModMatch(flowMod), tc.expectedMatcherStr)
+		}
+	})
+	t.Run("MatchARPSpa/MatchARPTpa", func(t *testing.T) {
+		testCases := []struct {
+			isSrc              bool
+			ip                 net.IP
+			expectedMatcherStr string
+			expectedMatchField *openflow15.MatchField
+		}{
+			{
+				isSrc:              true,
+				ip:                 net.ParseIP("1.1.1.1"),
+				expectedMatcherStr: "arp_spa=1.1.1.1",
+				expectedMatchField: &openflow15.MatchField{
+					Class: openflow15.OXM_CLASS_OPENFLOW_BASIC,
+					Field: openflow15.OXM_FIELD_ARP_SPA,
+					Value: &openflow15.ArpXPaField{ArpPa: net.ParseIP("1.1.1.1")},
+				},
+			},
+			{
+				ip:                 net.ParseIP("1.1.1.1"),
+				expectedMatcherStr: "arp_tpa=1.1.1.1",
+				expectedMatchField: &openflow15.MatchField{
+					Class: openflow15.OXM_CLASS_OPENFLOW_BASIC,
+					Field: openflow15.OXM_FIELD_ARP_TPA,
+					Value: &openflow15.ArpXPaField{ArpPa: net.ParseIP("1.1.1.1")},
+				},
+			},
+		}
+		for _, tc := range testCases {
+			var fb *ofFlowBuilder
+			if tc.isSrc {
+				fb = table.BuildFlow(1).MatchARPSpa(tc.ip).(*ofFlowBuilder)
+				assert.Equal(t, tc.ip, *fb.ofFlow.Match.ArpSpa)
+			} else {
+				fb = table.BuildFlow(1).MatchARPTpa(tc.ip).(*ofFlowBuilder)
+				assert.Equal(t, tc.ip, *fb.ofFlow.Match.ArpTpa)
+			}
+
+			flowMod := getFlowMod(t, fb.Done())
+			assert.Equal(t, 1, len(flowMod.Match.Fields))
+			checkMatchField(t, tc.expectedMatchField, &flowMod.Match.Fields[0])
+			assert.Contains(t, getFlowModMatch(flowMod), tc.expectedMatcherStr)
+		}
+	})
+	t.Run("MatchARPOp", func(t *testing.T) {
+		op := uint16(1)
+		expectedMatcherStr := "arp_op=1"
+		expectedMatchField := &openflow15.MatchField{
+			Class: openflow15.OXM_CLASS_OPENFLOW_BASIC,
+			Field: openflow15.OXM_FIELD_ARP_OP,
+			Value: &openflow15.ArpOperField{ArpOper: 1},
+		}
+
+		fb := table.BuildFlow(1).MatchARPOp(op).(*ofFlowBuilder)
+		assert.Equal(t, op, fb.ofFlow.Match.ArpOper)
+
+		flowMod := getFlowMod(t, fb.Done())
+		assert.Equal(t, 1, len(flowMod.Match.Fields))
+		checkMatchField(t, expectedMatchField, &flowMod.Match.Fields[0])
+		assert.Contains(t, getFlowModMatch(flowMod), expectedMatcherStr)
+
+	})
+	t.Run("MatchIPDSCP", func(t *testing.T) {
+		dscp := uint8(1)
+		expectedMatcherStr := "ip_dscp=1"
+		expectedMatchField := &openflow15.MatchField{
+			Class: openflow15.OXM_CLASS_OPENFLOW_BASIC,
+			Field: openflow15.OXM_FIELD_IP_DSCP,
+			Value: &openflow15.IpDscpField{Dscp: 1},
+		}
+
+		fb := table.BuildFlow(1).MatchIPDSCP(dscp).(*ofFlowBuilder)
+		assert.Equal(t, dscp, fb.ofFlow.Match.IpDscp)
+
+		flowMod := getFlowMod(t, fb.Done())
+		assert.Equal(t, 1, len(flowMod.Match.Fields))
+		checkMatchField(t, expectedMatchField, &flowMod.Match.Fields[0])
+		assert.Contains(t, getFlowModMatch(flowMod), expectedMatcherStr)
```

Comment:
I feel many subtests can be merged into table-driven one, by using function parameter like above. Then the input includes the builder function to call, the expected match field, and the expected match string. Many redundancy can be reduced.

Reply:
Done